hazelcast-client-new/src/main/java/com/hazelcast/client/proxy/ClientListProxy.java

```diff
@@ -325,28 +285,59 @@ public int indexOf(Object o) {
         ClientMessage request = ListListIteratorCodec.encodeRequest(name, index);
         ClientMessage response = invoke(request);
         ListListIteratorCodec.ResponseParameters resultParameters = ListListIteratorCodec.decodeResponse(response);
-        Collection<Data> resultCollection = resultParameters.list;
-        final List<E> list = new ArrayList<E>(resultCollection.size());
-        for (Data value : resultCollection) {
-            list.add((E) toObject(value));
-        }
-        return list.listIterator();
+        List<Data> resultCollection = (List<Data>) resultParameters.list;
+        SerializationService serializationService = getContext().getSerializationService();
+        return new UnmodifiableLazyList<E>(resultCollection, serializationService).listIterator();
     }
 
     public List<E> subList(int fromIndex, int toIndex) {
         ClientMessage request = ListSubCodec.encodeRequest(name, fromIndex, toIndex);
         ClientMessage response = invoke(request);
         ListSubCodec.ResponseParameters resultParameters = ListSubCodec.decodeResponse(response);
-        Collection<Data> resultCollection = resultParameters.list;
-        final List<E> list = new ArrayList<E>(resultCollection.size());
-        for (Data value : resultCollection) {
-            list.add((E) toObject(value));
-        }
-        return list;
+        List<Data> resultCollection = (List<Data>) resultParameters.list;
+        SerializationService serializationService = getContext().getSerializationService();
+        return new UnmodifiableLazyList<E>(resultCollection, serializationService);
```

Comment:
Previous return type `ArrayList` was `Serializable` but new `UnmodifiableLazyList ` is not. May this be a problem @gurbuzali WDYT?